File: stdlib/print.ml

module Print = struct

  val c_int: int #-> unit = "print_int"
  val c_newline: unit #-> unit = "print_newline"
  val c_float: float #-> unit = "print_float"
  val c_string: string #-> unit = "print_string"

  val int: int -> unit
  let int x = c_int x

  val newline: unit -> unit
  let newline () = c_newline()

  val float: float -> unit
  let float x = c_float x

  val string: string -> unit
  let string x = c_string x
end

File: stdlib/print.c
#include"liml.h"
#include<stdio.h>

lvalue print_int(lvalue n__){
  lint n = (lint) n__ ;

#ifdef ARCH_32
  printf("%d", n) ;
#endif

#ifdef ARCH_64
  printf("%ld", n) ;
#endif

  return 0 ;
}

lvalue print_newline(){
  printf("\n") ;
  return 0 ;
}

lvalue print_float(lvalue x__){
  lfloat x = V2F(x__) ;
  printf("%f", x) ;
  return 0 ;
}

lvalue print_string(lvalue x__){
  char *x = (char*) x__ ;
  printf("%s", x) ;
  return 0 ;
}

lvalue print_char(lvalue x){
  printf("%c", (char)x) ;
  return 0;
}
